File: AmsterdamEats.Maui/AmsterdamEats.Maui/ViewModels/ResultsViewModel.cs
using System.Collections.ObjectModel;
using CommunityToolkit.Mvvm.ComponentModel;
using AmsterdamEats.Models;
using AmsterdamEats.Services;

namespace AmsterdamEats.ViewModels;

public partial class ResultsViewModel : ObservableObject
{
    private readonly GooglePlacesService _placesService;
    private readonly ZenchefService _zenchefService;

    public ResultsViewModel(GooglePlacesService placesService, ZenchefService zenchefService)
    {
        _placesService = placesService;
        _zenchefService = zenchefService;
    }

    [ObservableProperty]
    private ObservableCollection<Restaurant> _restaurants = [];

    [ObservableProperty]
    private bool _isLoading;

    [ObservableProperty]
    [NotifyPropertyChangedFor(nameof(HasError))]
    private string? _errorMessage;

    public bool HasError => !string.IsNullOrEmpty(ErrorMessage);

    public async Task LoadAsync(SearchCriteria criteria)
    {
        IsLoading = true;
        ErrorMessage = null;
        Restaurants = [];

        try
        {
            var results = await _placesService.SearchRestaurantsAsync(criteria);
            Restaurants = new ObservableCollection<Restaurant>(results);

            // Check all restaurants' Zenchef availability in parallel
            var tasks = results.Select(async r =>
            {
                var rid = await _zenchefService.ResolveRestaurantIdAsync(r);
                var slots = rid.HasValue
                    ? await _zenchefService.CheckAvailabilityAsync(rid.Value, criteria)
                    : [];
                return (Restaurant: r, Rid: rid, Slots: slots);
            });

            foreach (var (restaurant, rid, slots) in await Task.WhenAll(tasks))
            {
                restaurant.ZenchefRestaurantId = rid;
                restaurant.AvailableSlots = slots;
                restaurant.AvailabilityChecked = true;
            }

            // Available + bookable first, then by Google rating
            Restaurants = new ObservableCollection<Restaurant>(
                results.OrderByDescending(r => r.HasAvailability).ThenByDescending(r => r.Rating));
        }
        catch (Exception ex)
        {
            ErrorMessage = ex.Message;
        }
        finally
        {
            IsLoading = false;
        }
    }
}

File: AmsterdamEats/AmsterdamEats/ViewModels/ResultsViewModel.swift
import Foundation

@MainActor
final class ResultsViewModel: ObservableObject {
    enum ViewState {
        case idle
        case loading
        case loaded([Restaurant])
        case error(String)
    }

    @Published var state: ViewState = .idle

    private let placesService = GooglePlacesService()
    private let zenchefService = ZenchefService()

    func load(criteria: SearchCriteria) async {
        state = .loading
        do {
            var restaurants = try await placesService.searchRestaurants(criteria: criteria)
            state = .loaded(restaurants)

            // Check Zenchef availability for all restaurants in parallel
            await withTaskGroup(of: (Int, Int?, [ZenchefSlot]).self) { group in
                for (index, restaurant) in restaurants.enumerated() {
                    group.addTask {
                        let rid = await self.zenchefService.resolveRestaurantId(for: restaurant)
                        var slots: [ZenchefSlot] = []
                        if let rid {
                            slots = (try? await self.zenchefService.checkAvailability(
                                restaurantId: rid, criteria: criteria)) ?? []
                        }
                        return (index, rid, slots)
                    }
                }
                for await (index, rid, slots) in group {
                    restaurants[index].zenchefRestaurantId = rid
                    restaurants[index].availableSlots = slots
                    restaurants[index].availabilityChecked = true
                }
            }

            // Available + bookable first, then by Google rating
            restaurants.sort {
                if $0.hasAvailability != $1.hasAvailability { return $0.hasAvailability }
                return $0.rating > $1.rating
            }
            state = .loaded(restaurants)
        } catch {
            state = .error(error.localizedDescription)
        }
    }
}
